pkg/k8shandler/elasticsearch.go

```diff
@@ -0,0 +1,88 @@
+package k8shandler
+
+import (
+	api "github.com/openshift/elasticsearch-operator/pkg/apis/logging/v1"
+	"github.com/sirupsen/logrus"
+)
+
+// this function should be called before we try doing operations to make sure all our nodes are
+// first ready
+func (er *ElasticsearchRequest) ClusterReady() bool {
+	// bypass using Status -- check all our pods first, make sure they're all 'Ready'
+	podStates := er.GetCurrentPodStateMap()
+
+	for _, stateMap := range podStates {
+		if len(stateMap[api.PodStateTypeNotReady]) > 0 || len(stateMap[api.PodStateTypeFailed]) > 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// this function should be called before we try doing operations to make sure any of our nodes
+// are first ready
+func (er *ElasticsearchRequest) AnyNodeReady() bool {
+	podStates := er.GetCurrentPodStateMap()
+
+	for _, stateMap := range podStates {
+		if len(stateMap[api.PodStateTypeReady]) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (er *ElasticsearchRequest) updateMinMasters() {
+	// do as best effort -- whenever we create a node update min masters (if required)
+	if er.AnyNodeReady() {
+		cluster := er.cluster
+		esClient := er.esClient
+
+		currentMasterCount, err := esClient.GetMinMasterNodes()
+		if err != nil {
+			logrus.Debugf("Unable to get current min master count for cluster %q in namespace %q", cluster.Name, cluster.Namespace)
+		}
+
+		desiredMasterCount := getMasterCount(cluster)/2 + 1
+		currentNodeCount, err := esClient.GetClusterNodeCount()
+		if err != nil {
+			logrus.Warnf("Unable to get cluster node count for cluster %q in namespace %q: %s", cluster.Name, cluster.Namespace, err.Error())
+		}
+
+		// check that we have the required number of master nodes in the cluster...
+		if currentNodeCount >= desiredMasterCount {
+			if currentMasterCount != desiredMasterCount {
+				if _, setErr := esClient.SetMinMasterNodes(desiredMasterCount); setErr != nil {
+					logrus.Debugf("Unable to set min master count to %d for cluster %q in namespace %q", desiredMasterCount, cluster.Name, cluster.Namespace)
+				}
+			}
+		}
+	}
+}
+
+func (er *ElasticsearchRequest) ensureAllShardAllocation() {
```

Comment:
`ensure` method should return error if cannot succeed, and should retry before returning failure?

Reply:
maybe this should be renamed then... given the nature of the reconcile loop there isn't a need to retry in the function and we don't want to block if for some reason we can't communicate with the cluster (e.g. we may need to do a cert full cluster restart)